controllers/grpcroute_controller.go

```diff
@@ -0,0 +1,376 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gateway_api_v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gateway_api_v1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	mcs_api "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/aws/aws-application-networking-k8s/controllers/eventhandlers"
+	"github.com/aws/aws-application-networking-k8s/pkg/aws"
+	"github.com/aws/aws-application-networking-k8s/pkg/config"
+	"github.com/aws/aws-application-networking-k8s/pkg/deploy"
+	"github.com/aws/aws-application-networking-k8s/pkg/deploy/lattice"
+	"github.com/aws/aws-application-networking-k8s/pkg/gateway"
+	"github.com/aws/aws-application-networking-k8s/pkg/k8s"
+	"github.com/aws/aws-application-networking-k8s/pkg/latticestore"
+	"github.com/aws/aws-application-networking-k8s/pkg/model/core"
+	latticemodel "github.com/aws/aws-application-networking-k8s/pkg/model/lattice"
+	lattice_runtime "github.com/aws/aws-application-networking-k8s/pkg/runtime"
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+// GRPCRouteReconciler reconciles a GRPCRoute object
+type GRPCRouteReconciler struct {
+	client           client.Client
+	scheme           *runtime.Scheme
+	finalizerManager k8s.FinalizerManager
+	eventRecorder    record.EventRecorder
+	modelBuilder     gateway.LatticeServiceBuilder
+	stackDeployer    deploy.StackDeployer
+	latticeDataStore *latticestore.LatticeDataStore
+	stackMashaller   deploy.StackMarshaller
+}
```

Comment:
Can you reuse http reconciler, dependencies are same, you can add conditional check when reconcile event comes, then branch into http/grpc. 
You may find some methods in controller are generic to route handling, where new route abstraction can help.

Reply:
The conditional check would look like this
```go
func (r *RouteReconciler) getRoute(ctx context.Context, req ctrl.Request) (core.Route, error) {
	var err error

	grpcRoute := &core.GRPCRoute{}
	if err = r.client.Get(ctx, req.NamespacedName, grpcRoute.K8sObject()); err == nil {
		return grpcRoute, nil
	}

	httpRoute := &core.HTTPRoute{}
	if err = r.client.Get(ctx, req.NamespacedName, httpRoute.K8sObject()); err == nil {
		return httpRoute, nil
	}

	return nil, err
}
```

My concern here is that this relies on the name and namespace of the resource. If I create an HTTPRoute with the same name and namespace as a GRPCRoute, this check cannot tell the difference between which one is being reconciled. As a result, I would rather opt for two separate controllers, but reuse some common functions between them.

Reply:
Yup, I have a working consolidated controller and event handler for routes. Will include it in next commit!